cmake/compiler/armclang/target.cmake

```diff
@@ -64,3 +64,7 @@ if(CONFIG_ARMCLANG_STD_LIBC)
   # library, for example error numbers, errno.h.
   list(APPEND TOOLCHAIN_C_FLAGS -D_AEABI_PORTABILITY_LEVEL=1)
 endif()
+
+if(CONFIG_CPP)
+  list(APPEND EXTRA_CXXFLAGS "-xc++")
```

Comment:
afaict this will cause all c files to build a c++ files.
This is different from what we do with gcc, where `.c` files are compiled using gcc and `.cpp` files are compiled using g++.

Only in specific cases like here:
https://github.com/zephyrproject-rtos/zephyr/blob/018bdf50227fee26ad3adc2251e53e0600889af2/tests/subsys/logging/log_api/CMakeLists.txt#L10-L13
we request specific C files to be compiled as `C++`, but not everywhere.

Is the reason for this PR that this principle fails with armclang ?
If not, can you elaborate more on the errors seen.

Reply:
EXTRA_CXXFLAGS is only added to C++:

From `cmake/extra_flags.cmake `:
```
if(EXTRA_CXXFLAGS)
  foreach(F ${EXTRA_CXXFLAGS_AS_LIST})
    zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:${F}>)
  endforeach()
endif()
```

Reply:
For GNU:

`GNU-CXX.cmake:set(CMAKE_CXX_COMPILE_OPTIONS_EXPLICIT_LANGUAGE -x c++)`

For `CMAKE_CXX_COMPILE_OPTIONS_EXPLICIT_LANGUAGE` (or `CMAKE_C_COMPILE_OPTIONS_EXPLICIT_LANGUAGE`)  is not set for ARMClang

Reply:
seems CMake doesn't honor the `LANGUAGE CXX` property.

Need to investigate if Zephyr toolchain infrastructure is the root cause or CMake itself.

Reply:
Looks like cmake itself.  In doing a trace.


```
/usr/share/cmake-3.22/Modules/CMakeCXXInformation.cmake(25):  include(Compiler/ARMClang-CXX OPTIONAL )
/usr/share/cmake-3.22/Modules/Compiler/ARMClang-CXX.cmake(1):  include(Compiler/Clang-CXX )
/usr/share/cmake-3.22/Modules/Compiler/Clang-CXX.cmake(1):  include(Compiler/Clang )
```

However we never see `CMAKE_CXX_COMPILER_FRONTEND_VARIANT` set to `xGNU` anywhere.

Reply:
@tejlmand so I think its a limitation of `cmake/Modules/CMakeDetermineCompilerId.cmake`

For arm-clang CMAKE_CXX_COMPILER_ID = `ARMClang` and thus the code in `CMakeDetermineCompilerId.cmake` ends up setting `CMAKE_CXX_COMPILER_FRONTEND_VARIANT` to "".  Where clang has a test https://github.com/Kitware/CMake/blob/c648d7363d1aaafce19ed0a44331a0925db9a8e6/Modules/CMakeDetermineCompilerId.cmake#L231